File: app/src/main/java/com/lzm/lightLive/util/SpUtils.java
package com.lzm.lightLive.util;

import android.content.Context;
import android.content.SharedPreferences;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.Map;

public class SpUtils {

    private static final String TAG = "mSp_u8xixAoiStrWdfW";

    private static SpUtils spUtils = null;

    private static SharedPreferences sp = null;

    private SpUtils(Context context) {
        sp = context.getSharedPreferences(TAG, Context.MODE_PRIVATE);
    }

    public static synchronized void initSpUtil(Context context) {
        if(spUtils == null) {
            spUtils = new SpUtils(context);
        }
    }

    public static synchronized SpUtils getInstance(Context context) {
        if (null == spUtils) {
            spUtils = new SpUtils(context);
        }
        return spUtils;
    }

    public void put(String key, Object object) {
        SharedPreferences.Editor editor = sp.edit();
        if (object instanceof String) {
            editor.putString(key, (String) object);
        } else if (object instanceof Integer) {
            editor.putInt(key, (Integer) object);
        } else if (object instanceof Boolean) {
            editor.putBoolean(key, (Boolean) object);
        } else if (object instanceof Float) {
            editor.putFloat(key, (Float) object);
        } else if (object instanceof Long) {
            editor.putLong(key, (Long) object);
        } else {
            editor.putString(key, object.toString());
        }
        SharedPreferencesCompat.apply(editor);
    }

    public void putString(String key, String stringData) {
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(key, stringData);
        editor.apply();
    }

    public Object get(String key, Object defaultObject) {
        if (defaultObject instanceof String) {
            return sp.getString(key, (String) defaultObject);
        } else if (defaultObject instanceof Integer) {
            return sp.getInt(key, (Integer) defaultObject);
        } else if (defaultObject instanceof Boolean) {
            return sp.getBoolean(key, (Boolean) defaultObject);
        } else if (defaultObject instanceof Float) {
            return sp.getFloat(key, (Float) defaultObject);
        } else if (defaultObject instanceof Long) {
            return sp.getLong(key, (Long) defaultObject);
        }
        return null;
    }

    public String getString(String key, String stringDataDefault) {
        return sp.getString(key, stringDataDefault);
    }

    public void remove(String key) {
        SharedPreferences.Editor editor = sp.edit();
        editor.remove(key);
        SharedPreferencesCompat.apply(editor);
    }

    public void clear() {
        SharedPreferences.Editor editor = sp.edit();
        editor.clear();
        SharedPreferencesCompat.apply(editor);
    }

    public boolean contains(String key) {
        return sp.contains(key);
    }

    public Map<String, ?> getAll() {
        return sp.getAll();
    }

    /**
     * 创建一个解决SharedPreferencesCompat.apply方法的一个兼容类
     */
    private static class SharedPreferencesCompat {
        private static final Method sApplyMethod = findApplyMethod();

        /**
         * 反射查找apply的方法
         */
        @SuppressWarnings({"unchecked", "rawtypes"})
        private static Method findApplyMethod() {
            try {
                Class clz = SharedPreferences.Editor.class;
                return clz.getMethod("apply");
            } catch (NoSuchMethodException e) {
            }

            return null;
        }

        public static void apply(SharedPreferences.Editor editor) {
            try {
                if (sApplyMethod != null) {
                    sApplyMethod.invoke(editor);
                    return;
                }
            } catch (IllegalArgumentException e) {
            } catch (IllegalAccessException e) {
            } catch (InvocationTargetException e) {
            }
            editor.commit();
        }
    }


}

File: app/src/main/java/com/lzm/lightLive/util/SpUtils.kt
package com.lzm.lightLive.util

import android.content.Context
import android.content.SharedPreferences
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method

class SpUtils private constructor(context: Context) {
    fun put(key: String?, obj: Any) {
        val editor = sp!!.edit()
        when (obj) {
            is String -> {
                editor.putString(key, obj)
            }
            is Int -> {
                editor.putInt(key, obj)
            }
            is Boolean -> {
                editor.putBoolean(key, obj)
            }
            is Float -> {
                editor.putFloat(key, obj)
            }
            is Long -> {
                editor.putLong(key, obj)
            }
            else -> {
                editor.putString(key, obj.toString())
            }
        }
        SharedPreferencesCompat.apply(editor)
    }

    fun putString(key: String?, stringData: String?) {
        val editor = sp!!.edit()
        editor.putString(key, stringData)
        editor.apply()
    }

    operator fun get(key: String?, defaultObject: Any?): Any? {
        when (defaultObject) {
            is String -> {
                return sp!!.getString(key, defaultObject as String?)
            }
            is Int -> {
                return sp!!.getInt(key, (defaultObject as Int?)!!)
            }
            is Boolean -> {
                return sp!!.getBoolean(key, (defaultObject as Boolean?)!!)
            }
            is Float -> {
                return sp!!.getFloat(key, (defaultObject as Float?)!!)
            }
            is Long -> {
                return sp!!.getLong(key, (defaultObject as Long?)!!)
            }
            else -> return null
        }
    }

    fun getString(key: String?, stringDataDefault: String?): String? {
        return sp!!.getString(key, stringDataDefault)
    }

    fun remove(key: String?) {
        val editor = sp!!.edit()
        editor.remove(key)
        SharedPreferencesCompat.apply(editor)
    }

    fun clear() {
        val editor = sp!!.edit()
        editor.clear()
        SharedPreferencesCompat.apply(editor)
    }

    operator fun contains(key: String?): Boolean {
        return sp!!.contains(key)
    }

    val all: Map<String, *>
        get() = sp!!.all

    /**
     * 创建一个解决SharedPreferencesCompat.apply方法的一个兼容类
     */
    private object SharedPreferencesCompat {
        private val sApplyMethod = findApplyMethod()

        /**
         * 反射查找apply的方法
         */
        private fun findApplyMethod(): Method? {
            try {
                val clz: Class<*> = SharedPreferences.Editor::class.java
                return clz.getMethod("apply")
            } catch (e: NoSuchMethodException) {
            }
            return null
        }

        fun apply(editor: SharedPreferences.Editor) {
            try {
                if (sApplyMethod != null) {
                    sApplyMethod.invoke(editor)
                    return
                }
            } catch (e: IllegalArgumentException) {
            } catch (e: IllegalAccessException) {
            } catch (e: InvocationTargetException) {
            }
            editor.commit()
        }
    }

    companion object {
        private const val TAG = "mSp_u8xixAoiStrWdfW"
        private var spUtils: SpUtils? = null
        private var sp: SharedPreferences? = null
        @Synchronized
        fun initSpUtil(context: Context) {
            if (spUtils == null) {
                spUtils = SpUtils(context)
            }
        }

        @Synchronized
        fun getInstance(context: Context): SpUtils? {
            if (null == spUtils) {
                spUtils = SpUtils(context)
            }
            return spUtils
        }
    }

    init {
        sp = context.getSharedPreferences(TAG, Context.MODE_PRIVATE)
    }
}